### PR TITLE
pybind11 bindings and build process

### PR DESCRIPTION
**Pull request by EelcoHoogendoorn**
Hi,

Not sure if you are still using this at CG, but I had an occasion to use it recently, but wanted compatibility with newer python versions. I decided it was a good occasion to give pybind11 a try, which seems to be preferred nowadays over boost.python; and indeed it was a relative breeze to get a compiling setup that (should) work across different python versions and OSes. Once I got it working once so far it has worked on the first try with any other config.

The only dangling bit here is the travis/docker stuff. I dont know if you still have an opinion about this at all; my suggestion would be to simplify the travis stuff to have some CI on the repo, and I will try and revive the conda-forge build; which should be a cookie-cutter recipe with the current approach, and rely on that for more comprehensive feedback on the robustness of the build. Building wheels is also supported btw.

If you are no longer invested and dont have time to review ill just maintain a fork on my github, and point the conda-forge recipe to that?

### CODE REVIEW

**Comment by Korijn** [comment]
Thanks for opening this PR! I've actually been itching to take a look at this myself. Pybind11 is pretty cool, huh? I'll have a good look later and review thoroughly.

File: conda-recipe/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+
+requirements:
+  build:
+    - {{ compiler('cxx') }}
+
+  host:
+    - python {{ python }}
+    - pip
+    - pybind11 >=2.6.0
+    - boost-cpp # this is uncompiled source dependency only, and does not imply run requirements
```
Comment (ivoflipse) [inline conda-recipe/meta.yaml:20]: This is no longer needed right?

File: pyopcode/api.cpp [hunk 2]
```diff
@@ -1,122 +1,201 @@
-#include <iostream>
+
+#include <stdexcept>
+
 #include <boost/cstdfloat.hpp>
```
Comment (ivoflipse) [inline pyopcode/api.cpp:4]: Ah, I thought you got rid of boost completely, but no

Comment (EelcoHoogendoorn) [reply pyopcode/api.cpp:4]: Yeah; using the header-only part of boost isnt so bad though (not the compilation/linking/distribution disaster that the rest of it is). And for some reason the cpp standard does not think its useful to provide people with float types of a guaranteed number of bytes ('it all depends on a million factors involving your compiler version and hardware' used to be a good thing back in the day I suppose), but thats definitely what we want here, and boost provides.

**Comment by EelcoHoogendoorn** [comment]
> Thanks for opening this PR! I've actually been itching to take a look at this myself. Pybind11 is pretty cool, huh? I'll have a good look later and review thoroughly.

Yeah its pretty cool. I really appreciate that they have working examples under CI, and put in the effort to abstract away all the OS specific stuff. Its almost like this is the 21th century or something. This implementation has taken this repo as a starting point https://github.com/pybind/python_example, and it did a great job of 'just working', mostly.

That being said the documentation, while alright for a cpp project and covering the basics, is still mostly of the 'us strongly typed programmers need no silly documentation all information is there in the types what additional questions could you possibly have' kind. So if you want to know if reshaping an array is supported be prepared to crtl-f in the source. I wrote some utils for validating assumptions about an input array, and returning the content of a vector without making a copy and with correct memory management. Youd think that would come batteries included, but mostly the answer to such questions appears to be no. 

Still, I consider those minor complaints. Ive been trying to use cpp code from python for about 10 years now, and the ease with which I actually got something working, on all python versions, OSes and compilers, without writing any hacky and brittle nonsense myself, was definitely a quantum leap forward.

**Comment by EelcoHoogendoorn** [comment]
Ah even better yet; thought it was needed to pin the runtime python major/minor version, but it seems that as the provided example implies, you really dont need that. I can build using any python 3 and the tests work on python 2 even. They should advertise that more prominently in the pybind docs! Only tested on OSX so far though. EDIT: confirmed with people who seem to know what they are talking about; despite pybind calling back into the python process, it does something automagical thats effectively like dynamic linking to the shared python binaries, but not really.

**Comment by Korijn** [comment]
My initial feedback for making this acceptable for us would be to drop the usage of conda. I propose:

* boost should be registered as a git submodule in the repo
* since pybind11 and numpy are available as wheels on pypi, they can be put in `pyproject.toml` (aren't they already?)

Pip/setuptools can install what it needs from that file and that takes care of all the build- and test-time dependencies.

In essence, I think this project doesn't need conda (or poetry (or pipenv)), just a `pyproject.toml` and `setup.py` should suffice, to be able to build wheels for all platforms.

Conda-forge can then contain a recipe to either build from scratch or repackage the wheel.

**Comment by EelcoHoogendoorn** [comment]
> My initial feedback for making this acceptable for us would be to drop the usage of conda. I propose:
> 
> * boost should be registered as a git submodule in the repo
> * since pybind11 and numpy are available as wheels on pypi, they can be put in `pyproject.toml` (aren't they already?)
> 
> Pip/setuptools can install what it needs from that file and that takes care of all the build- and test-time dependencies.
> 
> In essence, I think this project doesn't need conda (or poetry (or pipenv)), just a `pyproject.toml` and `setup.py` should suffice, to be able to build wheels for all platforms.
> 
> Conda-forge can then contain a recipe to either build from scratch or repackage the wheel.

Why not both, I wonder? I think having the conda recipe around is independently valuable, even if we get it to work on conda forge. If you want at any time to pick up development on the conda packaging side again, the conda forge recipe isnt the best place to start. I agree the environment.yml doesnt add much; though it does not cost much either.

Wrt to the building of wheels; I imagined it must be trivial since the `python -p pip install .` command seems to perform its install via an intermediate wheel; but I must say I know next to nothing about best practises of building wheels; the numpy runtime requirement isnt in there currently, it only has a build requirements section, where numpy does not belong. Perhaps I will get around to wrapping my head around all that sometime in the not too distant future; but if a well thought out wheel is a requirement for merging, perhaps it isnt in the cards.

File: pyopcode/api.cpp (hunk 2 above)
Comment (ivoflipse) [reply pyopcode/api.cpp:4]: I'm mostly concerned about the packaging. Is boost-cpp available on Pypi?

Comment (ivoflipse) [reply pyopcode/api.cpp:4]: Right, the conda-forge recipe gets it from https://dl.bintray.com/boostorg/release/{{ version }}/source/{{ filename }}

Hence Korijn's comment to include it via a submodule

**Comment by Korijn** [comment]
I will try my hand at it sometime soon, it seems very simple since you already figured out all the hard parts.

File: pyopcode/api.cpp (hunk 2 above)
Comment (EelcoHoogendoorn) [reply pyopcode/api.cpp:4]: Yeah, adding boost via a git submodule is definitely an option. There are probably other ways of getting a guaranteed float32 in cpp, but this is the one I happen to know. But getting rid of the dependency altogether may very well be possible too.

**Comment by Korijn** [comment]
I just got it to work (without conda) (on Windows) locally, I pushed a commit to your branch, updating `README.md` and `setup.py`. What are your thoughts?

File: pyopcode/api.cpp (hunk 2 above)
Comment (Korijn) [reply pyopcode/api.cpp:4]: I just downloaded the source tarball from the website for now (for the non-conda build), although a git submodule would be neater.

**Comment by EelcoHoogendoorn** [comment]
> I just got it to work (without conda) (on Windows) locally, I pushed a commit to your branch, updating `README.md` and `setup.py`. What are your thoughts?

Nice!

Running ```pip wheel . -w ./dist --no-deps``` on my mac gives:

``` 
Created wheel for pyopcode: filename=pyopcode-0.4.0-cp38-cp38-macosx_10_14_x86_64.whl size=224665 sha256=75e20a9a3601e8edf6b0973cf16de2bd54f51fd2bfdebeeb89e94c23766bfc6d
  Stored in directory: /private/var/folders/pt/b_4mkhnd377945qjz8p23hbr0000gn/T/pip-ephem-wheel-cache-my6x295k/wheels/22/a5/66/19feb101496a4bc74eb6914ff1c0a47b52f397500e700d7915
```

Two problems with that: first the indicated path does not actually seem to exist. wtf?

Secondly, is it a problem that the .whl references '38' in its filename? Indeed I am using py38 to build it; but in conda this signals a python version specific package; which apparently is not necessary. Is this wheel meant to be python version independent; or do we need to signal that explicitly somehow?

**Comment by ivoflipse** [comment]
I think such a tag doesn't exist outside of conda at the moment, I'll have to check whether we have to build it for each version or if you can repackage it somehow

**Comment by Korijn** [comment]
> I think such a tag doesn't exist outside of conda at the moment, I'll have to check whether we have to build it for each version or if you can repackage it somehow

No, it does exist. I can indicate that the wheel is universal (py2/3 compatible pure wheel) in setup.cfg.

The docs also mention that universal wheels should be pure python, so this probably also makes the wheel OS agnostic, which is not what we want. We should be able to force a platform tag with the `--platform` option. I also have a trick up my sleeve (learned when working on pygfx/wgpu-py) to force it in setup.py.

**Comment by Korijn** [comment]
Just a heads up for any onlookers; I'll be looking into this on our private repo first and will upstream the changes eventually here if I get everything to work.

**Comment by Korijn** [comment]
I got everything to work on azure pipelines and our internal repository, except for the conda part. I'll open a PR on the conda forge recipe but I won't maintain/support it beyond initial creation.

**Comment by Korijn** [comment]
See #14 for the local PR

Opened a PR for the conda-recipe here: https://github.com/conda-forge/pyopcode-feedstock/pull/24

**Pull request closed without merging**